internal/config/database.go

```diff
@@ -0,0 +1,145 @@
+package config
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// configuration keys
+	dbURL             = "db.url"
+	dbMigrationsPath  = "db.migrations.path"
+	dbMaxIdleConn     = "db.max_idle_conn"
+	dbMaxOpenConn     = "db.max_open_conn"
+	dbConnMaxLifetime = "db.conn_max_lifetime"
+	dbName            = "db.name"
+	dbUser            = "db.user"
+	dbPassword        = "db.password"
+	dbHost            = "db.host"
+	dbPort            = "db.port"
+	dbProtocol        = "db.protocol"
+
+	// database protocol enum
+	_ DatabaseProtocol = iota
+	// DatabaseSQLite ...
+	DatabaseSQLite
+	// DatabasePostgres ...
+	DatabasePostgres
+	// DatabaseMySQL ...
+	DatabaseMySQL
+)
+
+// DatabaseConfig contains fields, which configure the various relational database backends.
+//
+// Flipt currently supports SQLite, Postgres and MySQL backends.
+type DatabaseConfig struct {
+	MigrationsPath  string           `json:"migrationsPath,omitempty"`
+	URL             string           `json:"url,omitempty"`
+	MaxIdleConn     int              `json:"maxIdleConn,omitempty"`
+	MaxOpenConn     int              `json:"maxOpenConn,omitempty"`
+	ConnMaxLifetime time.Duration    `json:"connMaxLifetime,omitempty"`
+	Name            string           `json:"name,omitempty"`
+	User            string           `json:"user,omitempty"`
+	Password        string           `json:"password,omitempty"`
+	Host            string           `json:"host,omitempty"`
+	Port            int              `json:"port,omitempty"`
+	Protocol        DatabaseProtocol `json:"protocol,omitempty"`
+}
+
+func (c *DatabaseConfig) init() (warnings []string, _ error) {
+	// read in configuration via viper
+	if viper.IsSet(dbURL) {
+		c.URL = viper.GetString(dbURL)
+
+	} else if viper.IsSet(dbProtocol) || viper.IsSet(dbName) || viper.IsSet(dbUser) || viper.IsSet(dbPassword) || viper.IsSet(dbHost) || viper.IsSet(dbPort) {
+		c.URL = ""
+
+		if viper.IsSet(dbProtocol) {
```

Comment:
Outside of the scope of this PR, but I'd love to figure out if there was a way viper could handle this logic for us. This would prevent the need to do the `if viper.IsSet(x) { y = x }` pattern everywhere like we do now and would make adding new config values easier

Reply:
💯 agreed. This feels like boilerplate that could go away.

Reply:
Looks like some part of this could help https://github.com/spf13/viper#extracting-a-sub-tree